components/script/dom/htmlelement.rs

```
@@ -467,52 +469,90 @@ impl HTMLElementMethods for HTMLElement {
 
-        // Step 2.
-        let fragment = DocumentFragment::new(&document);
+        // Step 2: Let next be this's next sibling.
+        let next = node.GetNextSibling();
 
-        // Step 3. The given value is already named 'input'.
+        // Step 3: Let previous be this's previous sibling.
+        let previous = node.GetPreviousSibling();
 
-        // Step 4.
-        let mut position = input.chars().peekable();
+        // Step 4: Let fragment be the rendered text fragment for the given value given this's node
+        // document.
+        let fragment = self.rendered_text_fragment(input);
 
-        // Step 5.
-        let mut text = String::new();
+        // Step 5: If fragment has no children, then append a new Text node whose data is the empty
+        // string and node document is this's node document to fragment.
+        if fragment.upcast::<Node>().children_count() == 0 {
+            let text_node = Text::new(DOMString::from("".to_owned()), &document);
 
-        // Step 6.
-        while let Some(ch) = position.next() {
-            match ch {
-                '\u{000A}' | '\u{000D}' => {
-                    if ch == '\u{000D}' && position.peek() == Some(&'\u{000A}') {
-                        // a \r\n pair should only generate one <br>,
-                        // so just skip the \r.
-                        position.next();
-                    }
+            fragment
+                .upcast::<Node>()
+                .AppendChild(text_node.upcast())
+                .unwrap();
+        }
 
-                    if !text.is_empty() {
-                        append_text_node_to_fragment(&document, &fragment, text);
-                        text = String::new();
-                    }
+        // Step 6: Replace this with fragment within this's parent.
+        if let Err(error) = parent.ReplaceChild(fragment.upcast(), node) {
+            warn!("Failed to replace child node: {:?}", error);
```

Here you can return the error, since the spec throws.